src/Codec/Xlsx/Parser/Stream.hs

```diff
@@ -246,227 +160,242 @@ deriving via AllowThunksIn
-
--- | Tag zip entries with `PsFileTags`
---
--- Essentiall @.xlsx@ file is a zip archive with bunch of @.xml@s stored
--- inside. This conduit ensures that we recognise all the needed files to
--- work with later. All of the tags are pushed in a state monad.
-tagFilesC
-  :: ( HasPSFiles env
-     , MonadState env m
-     , MonadThrow m
-     )
-  => Maybe (Either ZipEntry BS.ByteString)
-  -> ConduitT (Either ZipEntry BS.ByteString) BS.ByteString m ()
-tagFilesC = \case
-  Just (Left zipEntry) -> do
-   let !filePath = either id Text.decodeUtf8 (zipEntryName zipEntry)
-   fileLens .= getFileTag filePath
-   await >>= tagFilesC
-  Just (Right fdata) -> do
-    -- strange behaviour indeed, why it does not await already yielded data?
-    yield fdata
-    await >>= tagFilesC
-  Nothing -> pure ()
-
--- | Parse a xml event
-parseFileC
-  :: ( MonadThrow m
-     , HasSheetState m
-     )
-  => ConduitT Event SheetItem m ()
-parseFileC = await >>= parseEventLoop
+-- | Run a series of actions on an Xlsx file
+runXlsxM :: MonadIO m => FilePath -> XlsxM a -> m a
+runXlsxM xlsxFile (XlsxM act) = do
+  env0 <- liftIO $ MkXlsxMState <$> newIORef Nothing <*> newIORef Nothing
+  Zip.withArchive xlsxFile $ runReaderT act env0
+
+liftZip :: Zip.ZipArchive a -> XlsxM a
+liftZip = XlsxM . ReaderT . const
+
+{-# SCC getOrParseSharedStrings #-}
+getOrParseSharedStrings :: XlsxM (V.Vector Text)
+getOrParseSharedStrings = do
+  sharedStringsRef <- asks _xs_shared_strings
+  mSharedStrings <- liftIO $ readIORef sharedStringsRef
+  case mSharedStrings of
+    Just strs -> pure strs
+    Nothing -> do
+      sharedStrsSel <- liftZip $ Zip.mkEntrySelector "xl/sharedStrings.xml"
+      let state0 = initialSharedString
+      byteSrc <- liftZip $ Zip.getEntrySource sharedStrsSel
+      st <- runExpat state0 byteSrc $ \evs -> forM_ evs $ \ev -> do
+        mTxt <- parseSharedString ev
+        for_ mTxt $ \txt ->
+          ss_list %= (`DL.snoc` txt)
+      let sharedStrings = V.fromList $ DL.toList $ _ss_list st
+      liftIO $ writeIORef sharedStringsRef $ Just sharedStrings
+      pure sharedStrings
+
+-- | Returns information about the workbook, found in
+-- xl/workbook.xml. The result is cached so the XML will only be
+-- decompressed and parsed once inside a larger XlsxM action.
+getWorkbookInfo :: XlsxM WorkbookInfo
+getWorkbookInfo = do
+  ref <- asks _xs_workbook_info
+  mInfo <- liftIO $ readIORef ref
+  case mInfo of
+    Just info -> pure info
+    Nothing -> do
+      sel <- liftZip $ Zip.mkEntrySelector "xl/workbook.xml"
+      src <- liftZip $ Zip.getEntrySource sel
+      sheets <- runExpat mempty src $ \evs -> forM_ evs $ \case
+        StartElement ("sheet" :: ByteString) attrs -> do
+          nm <- maybe (throwM MalformedWorkbook) pure $ lookup ("name" :: ByteString) attrs
+          sheetId <- maybe (throwM MalformedWorkbook) pure $ lookup "sheetId" attrs
+          sheetNum <- either (const $ throwM MalformedWorkbook) pure $ eitherDecimal sheetId
```

Comment:
Here the error gets squashed, the user won't know if the workbook is malformed due to the 'name' missing, the sheetId missing, or decimal parsing failing. I think it's more graceful if the reason for failure gets attached to the error.

Reply:
Good point, I'll fix this on master when I have a moment.